services/blockchain-app-registry/shared/utils/downloadRepository.js

```diff
@@ -180,48 +197,149 @@ const buildEventPayload = async (allFilesModified) => {
+
+		// Only process metadata files
+		if (!isMetadataFile(fileName)) return;
+
+		if (!(network in groupedFiles)) groupedFiles[network] = {};
+		if (!(appName in groupedFiles[network])) groupedFiles[network][appName] = [];
+		groupedFiles[network][appName].push(fileInfo);
+	});
+	return groupedFiles;
+};
+
+const getModifiedFileNames = (groupedFiles) => {
+	const fileNames = [];
+
+	Object.keys(groupedFiles).forEach(network => {
+		const appsInNetwork = groupedFiles[network];
+
+		Object.keys(appsInNetwork).forEach(appName => {
+			const appFiles = appsInNetwork[appName];
+			appFiles.forEach(file => fileNames.push(file.filename));
+		});
+	});
+	return fileNames;
+};
+
 const syncWithRemoteRepo = async () => {
 	try {
 		const dataDirectory = config.dataDir;
 		const appDirPath = path.join(dataDirectory, repo);
-
 		const { lastSyncedCommitHash, latestCommitHash } = await getCommitInfo();
 
+		// Skip if there is no new commit
 		if (lastSyncedCommitHash === latestCommitHash) {
-			logger.info('Database is already up-to-date');
-		} else {
-			const diffInfo = await getDiff(lastSyncedCommitHash, latestCommitHash);
-			// TODO: Add a check to filter out non-blockchain apps related files
-			const filesChanged = diffInfo.data.files.map(file => file.filename);
-
-			await BluebirdPromise.map(
-				filesChanged,
-				async file => {
-					const result = await getFileDownloadURL(file);
-					const filePath = path.join(appDirPath, file);
-					await downloadFile(result.data.download_url, filePath);
-					logger.debug(`Successfully downloaded: ${file}`);
-
-					if (file.endsWith(FILENAME.APP_JSON) || file.endsWith(FILENAME.NATIVETOKENS_JSON)) {
-						const [network, appName, filename] = file.split('/').slice(-3);
-						await indexMetadataFromFile(network, appName, filename);
-						logger.debug('Successfully updated the database with the latest changes');
-					}
-				},
-				{ concurrency: filesChanged.length },
-			);
-
-			await keyValueTable.set(KV_STORE_KEY.COMMIT_HASH_UNTIL_LAST_SYNC, latestCommitHash);
-
-			if (filesChanged.length) {
-				const eventPayload = await buildEventPayload(filesChanged);
-				Signals.get('metadataUpdated').dispatch(eventPayload);
-			}
+			logger.info('Database is already up-to-date.');
+			return;
+		}
+
+		const fileUpdatedStatuses = [
+			GITHUB_FILE_STATUS.REMOVED,
+			GITHUB_FILE_STATUS.MODIFIED,
+			GITHUB_FILE_STATUS.CHANGED,
+		];
+		const diffInfo = await getDiff(lastSyncedCommitHash, latestCommitHash);
+		const groupedFiles = groupFilesByNetworkAndApp(diffInfo.data.files);
+		const connection = await getDbConnection(MYSQL_ENDPOINT);
+
+		await BluebirdPromise.map(
+			Object.keys(groupedFiles),
+			async (networkName) => {
+				const appsInNetwork = groupedFiles[networkName];
+
+				await BluebirdPromise.map(
+					Object.keys(appsInNetwork),
+					async (appName) => {
+						const appFiles = appsInNetwork[appName];
+						const dbTrx = await startDbTransaction(connection);
```

Comment:
Let's use only one DB connction at the top level. Is there a reason because of which we a connection per network? Also, how do we handle the failure scenario (in case a transaction fails)?

Reply:
Discussed offline